src/validators/signing/tests/oracle_functions.py

```diff
@@ -0,0 +1,89 @@
+        exit_signature_recover_threshold: int,
+    ):
+        self.oracle_privkeys = oracle_privkeys
+        self.oracle_pubkeys = oracle_pubkeys
+        self.exit_signature_recover_threshold = exit_signature_recover_threshold
+
+    def verify_signature_shards(
+        self,
+        validator_pubkey: BLSPubkey,
+        validator_index: int,
+        fork: ConsensusFork,
+        exit_signature_shards: ExitSignatureShards,
+    ):
+        # Decrypt the signature shards using the oracle private keys
+        exit_signatures_decrypted = []
+        for oracle_privkey, exit_signature_shard in zip(
+            self.oracle_privkeys, exit_signature_shards.exit_signatures
+        ):
+            exit_signatures_decrypted.append(
+                BLSSignature(
+                    bytes.fromhex(
+                        ecies.decrypt(
+                            oracle_privkey.secret, bytes.fromhex(exit_signature_shard[2:])
+                        ).hex()
+                    )
+                )
+            )
+
+        # Verify the pubkey shares reconstruct into the full validator pubkey
+        validator_pubkey_shares = {
+            idx: BLSPubkey(bytes.fromhex(s[2:]))
+            for idx, s in enumerate(exit_signature_shards.public_keys)
+        }
+        aggregate_key = self.get_aggregate_key(validator_pubkey_shares)
+        assert aggregate_key == validator_pubkey
+
+        # Verify the signature (shards)
+        message = get_exit_message_signing_root(
+            validator_index=validator_index,
+            genesis_validators_root=settings.network_config.GENESIS_VALIDATORS_ROOT,
+            fork=fork,
+        )
+        for idx, (signature, validator_pubkey_share) in enumerate(
+            zip(exit_signatures_decrypted, exit_signature_shards.public_keys)
+        ):
+            pubkey = bytes.fromhex(validator_pubkey_share[2:])
+            assert bls.Verify(pubkey, message, signature) is True
+
+        # Verify the full reconstructed signature
+        signatures = dict(enumerate(exit_signatures_decrypted))
+        random_indexes = random.sample(sorted(signatures), k=self.exit_signature_recover_threshold)
+        random_signature_subset = {k: v for k, v in signatures.items() if k in random_indexes}
+        reconstructed_full_signature = reconstruct_shared_bls_signature(random_signature_subset)
+        assert (
+            bls.Verify(aggregate_key, message, reconstructed_full_signature) is True
+        ), 'Unable to reconstruct full signature'
+
+    @staticmethod
+    def get_aggregate_key(keys) -> BLSPubkey:
+        r = b.Z1
+        for i, key in keys.items():
+            key_point = pubkey_to_G1(key)
+            coef = 1
+            for j in keys:
+                if j != i:
+                    coef = -coef * (j + 1) * prime_field_inv(i - j, b.curve_order) % b.curve_order
+            r = b.add(r, b.multiply(key_point, coef))
+        return G1_to_pubkey(r)
```

Comment:
I suggest to move hardcore math out of tests. Somewhere to validators/signing folder.

Reply:
This function is also used in v3-oracle. So maybe we will move it to sw-utils later.

Reply:
I moved the math functions to `src/validators/signing/key_shares.py`.

I don't have access to the v3-oracle repository but used the implementation from one of your other repositories, it's probably the same.